src/Swift.Runtime/src/Swift/SwiftString.cs

```diff
@@ -44,18 +48,17 @@ public void Dispose()
 
     protected virtual void Dispose(bool disposing)
     {
-        if (!_disposed)
+        if (!_payload.IsInvalid)
         {
-            var metadata = SwiftObjectHelper<SwiftString>.GetTypeMetadata();
-
             unsafe
             {
-                fixed (void* payload = &_payload)
+                _payload.SetMetadata(SwiftObjectHelper<SwiftString>.GetTypeMetadata());
```

Comment:
I find this really problematic. If I read this correctly then the SafeHandle destructor will not work (not actually release anything) if it's called directly by GC, since it won't have the metadata set. It will only work if somebody calls Dispose on the wrapper (`SwiftString`).
That seems to defeat the whole purpose of this setup - for example as-is if I don't call `Dispose` on the string I don't see how the code will ever call the VWT Destroy.

Is this an optimization, where we don't want to construct the TypeMetadata before something calls Dispose? 
Honestly I was expecting the ctor for SwiftHandler to take the metadata since without it, it can't destroy the value.

Reply:
The GC shouldn't collect a SafeHandle without invoking the type's finalizer, which calls Dispose and sets the metadata. This PR doesn’t change the UX -- all projected types that require memory management have Dispose methods that set the metadata.

In general, you are right, I don't think this is the optimal approach and I didn’t want to make any decisions without discussing it here first. Here are some options:
 - Set the metadata in the constructor
 - Set the destroy function pointer in the constructor
 - Set a callback to a static method on the type, which will retrieve the metadata and call Destroy

Reply:
Personal preference: I don't like relying on the order of finalizers even though you're correct. One it's not obvious why it works, so at least some comments to that end should be added. Second, it feels fragile. Third, the usage of SafeHandle is really weird as is.

So for this I would go with some of the options you listed where the safe handle has a way to ask for the metadata, either through some callback or preinitialized. Personally I would probably start with preinitializing - and see if the perf hit is even noticable (we have pretty good caching of the metadata, so really only the 1st call should be noticable).

Reply:
Can a SafeHandle subclass be generic?

```cs
public sealed class SwiftHandle<T> : SafeHandleZeroOrMinusOneIsInvalid
	where T: ISwiftObject
{
    protected override unsafe bool ReleaseHandle()
    {
        var metadata = SwiftObjectHelper<T>.GetTypeMetadata());
        metadata.ValueWitnessTable->Destroy((void*)handle, _metadata);
        handle = IntPtr.Zero;
        return true;
    }
}
```

And then you'd do something like this:

```cs
public class SwiftString : IDisposable, ISwiftObject {
	private SwiftHandle<SwiftString> _payload;
}
```

That would avoid the problem of having to set the metadata.

Reply:
> Personal preference: I don't like relying on the order of finalizers even though you're correct. One it's not obvious why it works, so at least some comments to that end should be added. Second, it feels fragile. Third, the usage of SafeHandle is really weird as is.

Updated. The payload is now allocated on the C# side as native memory so the GC won't move it. The type's disposal is now decoupled from the native handle.

Reply:
> Can a SafeHandle subclass be generic?

Updated: SwiftHandle is now generic to avoid setting metadata in the initializer, while SafeHandle is used at the P/Invoke boundary.